### ISSUE TITLE
kotlin type DSL accepts Message or Builder to build from

### ISSUE THREAD

**Issue by Clement-Jean** [protocolbuffers/protobuf#9972]
**What language does this apply to?**

This applied to Kotlin generated code

**Describe the problem you are trying to solve.**

In the java generated code we are able to build from another Message or Builder by doing:

```newBuilder(messageOrBuilder)```

this will populate the existing fields to the new Builder.

**Describe the solution you'd like**

I think we simply need to generate an overload function of the Message Builder and add a parameter to it before the Builder lambda.

**Additional context**

I'd be happy to contribute to that if this gets accepted. This is quite time consuming to map all the fields by hand.

**Comment by elharo** [comment]
Sounds reasonable. It would require a Tier 2 review.

**Comment by deannagarcia** [comment]
I don't think I fully understand what you are proposing. Can you provide an example of what you have to do now to workaround this and what you would like the syntax to look like?

**Comment by Clement-Jean** [comment]
Right. Sorry for not being clear. Let's say I have the following message:

```
message SearchRequest {
  string query = 1;
  uint32 page = 2;
  uint32 page_size = 3;
}
```
and let's say that I already such a request but I need to create a new one and change the page_size. With the Kotlin DSL when creating such a request I need to do:

```
searchRequest {
    this.query = that.query
    this.page = that.page
    this.page_size = newPageSize
}
```

in Java though I could do the following:

```
SearchRequest.newBuilder(that).setPageSize(newPageSize).build()
```

I'm not reassigning all the fields.

So I hope we can generate something like:

```
public inline fun searchRequest(sr: SearchRequest, block: SearchRequestKt.Dsl.() -> kotlin.Unit): SearchRequest
```

and that instead of using the `newBuilder` in the `_create` function, we use the `sr.toBuilder()`. This would then assign all the already existing fields to the instance and we could only change the necessary ones.

Hope that's clearer

**Comment by deannagarcia** [comment]
Thanks for the example!

Kotlin protos are based on Java protos so you also have access to all Java protobuf methods, including the `newBuilder` that you want to use. So currently the easiest way to do what you are asking for is:
```
SearchRequest.newBuilder(that).apply {
    pageSize = newPageSize
}.build()
```

While this is better than the Kotlin DSL example you gave, I agree that the syntax you suggested is shorter, clearer, and more idiomatic. 

Given that we already have an implementation (even if worse) of this, I'm hesitant to prioritize the work. That being said, if you are still interested in contributing I can move forward with the team to see if we can get this proposal approved and ready to work on.

**Comment by Clement-Jean** [comment]
Thank you @deannagarcia , I'm interested in contributing to that yeah. I just might need little guidance since this is the first time I contribute to this repo.
